### ISSUE TITLE
🐛 [Bug Report] wotlk mangosd 100% CPU usage

### ISSUE THREAD

**Issue by xofyarg** [cmangos/issues#2358]
## 🐛 Bugreport
The wotlk mangosd process uses 100% CPU on one core, even the server is idle with no player online. The server and db are running inside docker containers. On the same host, classic runs without any problem, CPU usage is less than 10%.

### Version & Environment
Client Version: "3.3.5a" (WOTLK)

CMaNGOS Repo & Commit Hash:  [WOTLK](https://github.com/cmangos/mangos-wotlk/tree/01aa9ff863da7238793368f57b4b6e98f26359de)

Database Repo & Commit Hash:  [WOTLK](https://github.com/cmangos/wotlk-db/tree/3a41fd64419614bb3c332e29eb91fcbd4b7dc7e9)

Operating System: Linux

### Steps to reproduce
1. start mangosd
2. check CPU usage

### Debug info
```
$ ps -C mangosd -T -opid,tid,pcpu,comm
  PID   TID %CPU COMMAND
21726 21726  0.0 mangosd
21726 21733  0.4 mangosd
21726 21734  0.5 mangosd
21726 21735  0.4 mangosd
21726 21887 35.0 mangosd
21726 21888 35.3 mangosd
21726 21889 35.3 mangosd
21726 21890  0.6 mangosd
21726 21891  0.2 mangosd
21726 21892  0.0 mangosd
```

```
$ perf record -p 21726 -g sleep 10
$ perf report --stdio
# Total Lost Samples: 0
#
# Samples: 38K of event 'cycles:ppp'
# Event count (approx.): 6003585604
#
# Children      Self  Command  Shared Object        Symbol                                                                                               >
# ........  ........  .......  ...................  .....................................................................................................>
#
    20.08%     0.19%  mangosd  [unknown]            [.] 0000000000000000
            |
             --19.89%--0
                       |
                       |--2.75%--Unit::Update
                       |
                       |--1.13%--std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<cha>
                       |
                       |--0.69%--std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::b>
                       |
                       |--0.65%--WorldObject::UpdateCooldowns
                       |
                       |--0.55%--GameObject::Update
                       |
                       |--0.52%--std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::b>
                       |
                        --0.50%--MotionMaster::UpdateMotion

    12.35%     0.00%  mangosd  [unknown]            [.] 0x4955fc8949544100
            |
            ---0x4955fc8949544100
               Creature::~Creature
               0
               |
               |--2.72%--Unit::Update
               |
               |--0.67%--std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_str>
               |
                --0.65%--WorldObject::UpdateCooldowns

    12.35%     0.00%  mangosd  mangosd.debug        [.] Creature::~Creature
            |
            ---Creature::~Creature
               0
               |
               |--2.72%--Unit::Update
               |
               |--0.67%--std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_str>
               |
                --0.65%--WorldObject::UpdateCooldowns

     9.40%     8.89%  mangosd  libc-2.31.so         [.] malloc
            |
             --8.61%--malloc

     8.27%     0.00%  mangosd  [unknown]            [.] 0x0000000000000001
            |
            ---0x1
               |
               |--2.75%--0x7fae0e1fccd8
               |
                --0.52%--0x7fae0e1fc94f

     5.31%     0.00%  mangosd  [unknown]            [.] 0x0000441f0f000000
            |
            ---0x441f0f000000
               |
               |--4.51%--MapUpdateWorker::~MapUpdateWorker
               |          0x5651f7ca7858
               |          |
               |           --4.45%--MapUpdateWorker::execute
               |                     |
               |                      --3.35%--Map::Update
               |
                --0.80%--boost::any::holder<long>::~holder

     4.51%     0.00%  mangosd  mangosd.debug        [.] MapUpdateWorker::~MapUpdateWorker
            |
            ---MapUpdateWorker::~MapUpdateWorker
               0x5651f7ca7858
               |
                --4.45%--MapUpdateWorker::execute
...
```

**Comment by kelbren** [comment]
Can confirm, I get a constant 25-30% utilization when Idle. TBC core is typically around 3-5%

**Comment by jimmybrancaccio** [comment]
@xofyarg / @kelbren Are either of you using AH Bot?

**Comment by xofyarg** [comment]
No, only playerbot is enabled. Also maps are extracted using high resolution with mmaps, if that counts.

**Comment by jimmybrancaccio** [comment]
Ah okay, I must be running into a different issue. Have you tried with non-high res maps to see if that made any difference?

**Comment by kelbren** [comment]
I have both playerbot and ahbot options on during compile. AH works as expected.
Also have high res mmaps, tried both high and normal .. no noticeable change.

**Comment by jimmybrancaccio** [comment]
For what it's worth, I didn't have the hi-res maps (I do now), and had high CPU usage. I've been trying to dig into it with `strace` but haven't seen anything yet. I'll continue to try poking at it though...

![image](https://user-images.githubusercontent.com/405585/108724458-f49c8600-74ea-11eb-8709-0493db020bbe.jpeg)

**Comment by Pnumekin** [comment]
Not sure if it's the same problem, but the problem is now on classic too. One core is running 100% where it was not the case some month ago with same hardware/parameters.
Using classic core : 763e9b3cd92064b6eb0d43b1c3cd4ac3b7879688

**Comment by killerwife** [comment]
How many threads do you guys have allocated to the docker image. Also specs of cpu.

**Comment by Pnumekin** [comment]
Not using docker, should I open a new bug report ?
My cpu is [Intel Xeon CPU E3-1265L V2 @ 2.50GHz](https://www.intel.com/content/www/us/en/products/sku/65728/intel-xeon-processor-e31265l-v2-8m-cache-2-50-ghz/specifications.html)

**Comment by killerwife** [comment]
Well that is quite the potato. We used to run old 2018-2019 vengeance on that and with all the spectre mitigations and other performance drawbacks its pretty slow these days for singlethread performance.

**Comment by Pnumekin** [comment]
Ok it sounded strange to me because my cmangos build from one or two month ago is working perfectly fine with only 4% cpu usage. 
So I will stay on my build that is working :)

**Comment by killerwife** [comment]
Then figure out what commit did it. I can obviously be wrong.

**Comment by evil-at-wow** [comment]
> Well that is quite the potato.

It mostly depends on what you want to use it for. Obviously this ~10-year old CPU is not getting anywhere near a state of the art AMD Epyc 9374F or so, but for a small(er) server, it should be fine IMHO.

The use of docker shouldn't really matter either for CPU usage - I'm assuming memory is not severely limited or something like that - because the database and server are still normal processes running on the host ("docker is just a fancy name for user friendly use of control groups").

The perf output shown by xofyarg mostly hints to Update calls. That's not really a surprise because these do the actual work, but maybe some are running in some kind of "busy loop" for some reason?

Pnumekin: it seems you have a 'working' version from a while ago and a 'broken' version now; if you can spare the time and resources, would it be possible to do a git-bisect exercise to find the commit that seems to have broken it?

**Comment by Pnumekin** [comment]
Yes I will search to find the commit that changed cpu usage, just need some time to test all.

**Comment by killerwife** [comment]
Oh, come to think of it, since you are running on linux, are you compiling with -O2 ?

**Comment by killerwife** [comment]
And are you running War Effort or Scourge Invasion events?

**Comment by killerwife** [comment]
And as usual, fixed. Test it yourself again.

**Comment by evil-at-wow** [comment]
Well, well, well, would you look at that. A "busy loop". Called it :grin:

**Comment by killerwife** [comment]
So evil.

**Comment by Pnumekin** [comment]
@killerwife you're too fast, I didn't had time to git-bisect :smile: 
Compiled and tested, problem fixed !
(Have a stable 4% on my CPU)

**Comment by evil-at-wow** [comment]
I just realized that we (read: killerwife) fixed Classic, which is good news of course, but the original issue was logged against WotLK. So I checked and notice that ```LFGQueue::Update()``` isn't a busy loop in the WotLK core (it already has a similar 500ms sleep), so I guess there's still some other issue lurking in the WotLK core.

@xofyarg Any chance you can grab me a perf dump of about 30s when it's ~100% cpu and share the file somewhere for me to investigate what's causing this? Something like this should hopefully work: ```perf record -o mangosd-perf.data -p <pid> --call-graph dwarf sleep 30``` (or whatever file name you want besides mangosd-perf.data of course). I realize the resulting file may not be very small but hopefully you can share it somewhere (there are a few free file sharing services around if you don't have Google drive or something).

**Comment by xofyarg** [comment]
Hi @evil-at-wow, I'm not running mangos anymore unfortunately, not able to get a dump, sorry. I guess this is just another way to get rid of the issue :-/

**Comment by killerwife** [comment]
I did test wotlk when I was testing all cores, and from what I saw it was just progressively more utilized from vanilla through wotlk.

**Comment by evil-at-wow** [comment]
Fair enough @xofyarg, no problem at all! To be honest, I only just now realize that your original post was from early 2021... I just noticed the thread because of the recent activity :no_mouth:

killerwife: yes, that's the other thing I was thinking about yesterday. The LFG stuff is quite different in wotlk, simply because, well, it was something completely else in classic/tbc compared to (late) wotlk.

**Comment by killerwife** [comment]
Yeah and I mean, I added both systems, wotlk was written from 0 lines, vanilla was merged but altered from vmangos. Both are however in async threads, something that hasnt been done outside of networking.

**Comment by al3xc1985** [comment]
@insunaa wasn't this bug fixed through numerous leaking fixes by killerwife, you, evil-at-wow?

**Comment by evil-at-wow** [comment]
The memory leak fixes only fixed the server's memory usage growing over time, not its CPU use. So those were unrelated fixes.

Killerwife fixed a busy loop here (or actually, in the Classic core), but I don't think we know for sure that fixed all the high CPU usage. [Pnumekin](https://github.com/Pnumekin) reported it fixed the problem for Classic, but someone would need to check server idle CPU usage with a recent WoTLK core again. Alternatively, seeing no one else yelling here in over two years, we could close this issue assuming it is fixed, and then people can just open a new issue when they still have the problem?

**Comment by adellaci** [comment]
I am running my server with abbot and nothing else. I will report back

Feb 15, 2025 11:01:43 evil-at-wow ***@***.***>:

**Comment by adellaci** [comment]
I started Saturday early evening, the memory started at 2469M with CPU in the 20's. MySQL is allways very stable, never changes much, back and forth by 2M This screen is just now.
![Image](https://github.com/user-attachments/assets/2c96351b-c5e8-47ce-90f3-70855cb05e99)

**Comment by adellaci** [comment]
I will run it through to Saturday, and get you another snap shot. 
This is with;
Core e07c9c91900cc5fa7219920be13e9ca14693298e
DB 86308518c4879843989524d3f8dfa097f0221097

**Comment by adellaci** [comment]
@evil-at-wow, sorry. I am not running it in a docker container, this is a full blown server dedicated to wow, that will make a difference. Do you still want me to run it out till Saturday and still get a snapshot, will it help?

**Comment by al3xc1985** [comment]
This is very stable as I can see

**Comment by evil-at-wow** [comment]
Indeed, this seems perfectly acceptable. A few minutes of CPU time for the mangos process over the course of two days is nowhere near the 100% cpu usage reported originally. Don't worry about not running this in a docker container, for all practical purposes here it's the same thing as running it on bare metal (a docker container is essentially also just a set of processes running natively on the host, albeit "sandboxed" via control groups/kernel namespaces).

The memory usage may be a bit higher at ~4.8 GB, but if I remember correctly this is still in line with your previous reports like the end results in cmangos/issues#3140. It may even be perfectly normal if you reverted back to 'GridUnload = 0' and loading all grids.

Since this issue was mostly about CPU usage of the core, I think we can close this. If anyone still has this problem or it comes back one day, we can investigate in a new issue.

@adellaci As far as I'm concerned, there's no need to keep this running. Thank you very much for the efforts and reports! Much appreciated!

**Comment by adellaci** [comment]
I verified, i am running with these settings, GridUnload = 0, LoadAllGridsOnMaps = "".  When i started the server it was using 2469MB of memory. Though the memory leak is so much better there is still some. I will let it run so you can see what a week of running looks like, no one has logged in, and i will keep it that way, just to see. What i have also noticed is when the memory usage gets high, and i go to shutdown the server from the console, it take a relatively long time before it actually shuts down and the PIDs terminate.

**Comment by adellaci** [comment]
One week running, no one logged in 
![Image](https://github.com/user-attachments/assets/6661cc05-4102-4cf8-b714-29bdecf234f9)
